### PR TITLE
ENH: NetCDF format for storing and retrieving topographies, including metadata

### PR DESCRIPTION
**Pull request by pastewka**
This PR implements reading and writing of topographies into NetCDF files - also for line scans. We should be able to use this for caching, i.e. the topography object needs to be restored exactly as it was. (This only works if there is no pipeline attached to it.)

### CODE REVIEW

**Comment by pastewka** [comment]
@mcrot @AntoineSIMTEK - this PR now completely implements reading and writing of NetCDF files, also for line scans. It is implemented as the `to_netcdf` method. The `info` dictionary is written in JSON serialized form. What is written is the current state of the topograpgy (i.e. what you would get from `squeeze`) *not* the pipeline. My question is whether we want the pipeline in there:

Pros:
* You get exactly the same state as when writing
* May be used to store the pipeline in TopoBank

Cons:
* Difficult to read by someone else (other than `SurfaceTopography`) - but you can always do `.squeeze().to_netcdf('xxx.nc')`
* Whenever we change API (i.e. add an argument to an existing pipeline function), this may break reading of existing NetCDF files. We would need some sort of migration/change management.

I think the last point is a show stopper and we should not do this. What is your opinion?

**Comment by pastewka** [comment]
Another comment: Needing serialization of the `info` dictionary means it cannot contain `datetime` objects. I've switched the `acquisition_data` to a simple string.

**Comment by AntoineSIMTEK** [comment]
I am very happy that we now have a function that allows to store and load a topography __including__ physical_sizes with netcdf. 

I agree with you that storing the pipeline is complex and will make problems with different versions. I think we shouldn't do that.

**Comment by pastewka** [comment]
Yet another comment: I always thought that `physical_sizes` cannot be `None` for the topography (readers can return `None` but then ask you to specify it upon reading), but some readers actually returned topographies that were `None`. I lifted that requirement entirely, topographies no can have no `physical_sizes`. Are there any objections to this?

**Comment by AntoineSIMTEK** [comment]
This will blow up in some analysis functions.
I think we should change the readers so that they ask for the physical_sizes in the Topography call if there is nothing in the file.

**Comment by pastewka** [comment]
Okay, I will change that back. The problem is actually the legacy readers, that return None (and need to because there is no way yet to read the header.)

@mcrot - What are the implications of that for TopoBank?

**Comment by pastewka** [comment]
@AntoineSIMTEK - can you review this PR, I will open a new one for the `None` issue...

**Comment by pastewka** [comment]
See here #77

File: SurfaceTopography/IO/NC.py [hunk 1]
```diff
@@ -129,44 +136,91 @@ def __init__(self, fobj, communicator=None):
...
+                    ny = self._y_dim.size
+                except AttributeError:
+                    # scipy.io.netcdf_file
+                    ny = self._y_dim
+                nb_grid_pts = (nx, ny)
+        else:
+            # This is a nonuniform line scan
+            try:
+                # netCDF4
+                nx = self._x_dim.size
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/IO/NC.py:215]: Does _x_dim exist in the nonuniform scans ? Or what is the purpose of _n_dim ?

File: SurfaceTopography/IO/NC.py [hunk 2]
```diff
@@ -175,22 +229,43 @@ def topography(self, channel_index=None, physical_sizes=None,
...
-            return Topography(self._heights_var[...], physical_sizes,
-                              periodic=self._periodic
-                              if periodic is None else periodic,
+            if self._y_dim is None:
+                raise ValueError('Parallel reading only works for topographies, not line scans.')
+
+            physical_sizes = self._check_physical_sizes(physical_sizes,
+                                                        self._physical_sizes)
+
+            return Topography(np.array(self._heights_var[...]), physical_sizes,
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/IO/NC.py:267]: If you call self._heights_var[subdomain_slices], does it only read load the subdomain slice ?

File: SurfaceTopography/IO/NC.py [hunk 3]
```diff
@@ -230,33 +305,108 @@ def write_nc(topography, filename, format='NETCDF3_64BIT_OFFSET'):
...
-            y_var.length_unit = no_uft8_height_unit(topography.info['unit'])
-        y_var[...] = (np.arange(ny) + 0.5) * sy / ny
+            x_var = nc.createVariable('x', 'f8', ('x',))
+            x_var.length = sx
+
+            x_var.periodic = 1 if topography.is_periodic else 0
+            if 'unit' in topography.info:
+                # scipy.io.netcdf_file does not support UTF-8
+                x_var.unit = mangle_length_unit_ascii(topography.info['unit'])
+            x_var[...] = (np.arange(nx) + 0.5) * sx / nx
```
Comment (AntoineSIMTEK) [inline SurfaceTopography/IO/NC.py:346]: Should we be efficient and only store the physical sizes instead of the full array ?

Comment (AntoineSIMTEK) [inline SurfaceTopography/IO/NC.py:346]: The "+0.5" is not consistent with the positions method

**Review by AntoineSIMTEK — CHANGES_REQUESTED** [review]

File: SurfaceTopography/IO/NC.py (hunk 1 above)
Comment (pastewka) [reply SurfaceTopography/IO/NC.py:215]: No, nonuniform scans have only `_n_dim`. The idea is to distinguish between equally spaces sets `_x_dim` and nonequally spaces sets `_n_dim`, where `n` is the number of data points. I could be convinced to just call it `x_dim` in both cases.

File: SurfaceTopography/IO/NC.py (hunk 2 above)
Comment (pastewka) [reply SurfaceTopography/IO/NC.py:267]: Yes. I guess that is implementation specific, but I belive that both `netCDF4` and `scipy.io.netcdf` will just read that subsection. (In `scipy.io.netcdf` it is opened as a memory-mapped buffer anyway.)

File: SurfaceTopography/IO/NC.py (hunk 1 above)
Comment (AntoineSIMTEK) [reply SurfaceTopography/IO/NC.py:215]: But how do you detect uniform versus nonuniform ?

File: SurfaceTopography/IO/NC.py (hunk 3 above)
Comment (pastewka) [reply SurfaceTopography/IO/NC.py:346]: This is simply for convenience. If you open the NetCDF file in Python or MATLAB, you can just play heights vs x and that shows the topography. This works for nonuniform and uniform cases.

Comment (pastewka) [reply SurfaceTopography/IO/NC.py:346]: Ah, we should probably get this from the positions method in the first place. Good point.

Comment (AntoineSIMTEK) [reply SurfaceTopography/IO/NC.py:346]: Ok makes sense.

File: SurfaceTopography/IO/NC.py (hunk 1 above)
Comment (pastewka) [reply SurfaceTopography/IO/NC.py:215]: From the presence of the `length` attribute, but I guess that having a different name for the dimension is clearer.

Comment (pastewka) [reply SurfaceTopography/IO/NC.py:215]: That said, I need to check support for periodic nonuniform line scans which should actually need the length - but this seems to be broken in the code. We should probably simply not support periodic nonuniform line scans.

Comment (pastewka) [reply SurfaceTopography/IO/NC.py:215]: See #78

**Comment by mcrot** [comment]
> Okay, I will change that back. The problem is actually the legacy readers, that return None (and need to because there is no way yet to read the header.)
> 
> @mcrot - What are the implications of that for TopoBank?

Currently in TopoBank, there is some code which catches `None` in various variants and uses then
the values entered by the user in last upload form:

```python
# Set size if physical size was not given in datafile
            # (see also  TopographyCreateWizard.get_form_initial)
            # Physical size is always a tuple or None.
            channel_dict = toporeader.channels[self.data_source]
            channel_physical_sizes = channel_dict.physical_sizes
            physical_sizes_is_None = channel_physical_sizes is None \
                                     or (channel_physical_sizes == (None,)) \
                                     or (channel_physical_sizes == (None, None))
            # workaround, see GH 299 in Pyco

            if physical_sizes_is_None:
                if self.size_y is None:
                    topography_kwargs['physical_sizes'] = self.size_x,
                else:
                    topography_kwargs['physical_sizes'] = self.size_x, self.size_y
```
The dict `topography_kwargs` is finally passed to the reader. So `None`should be no problem.

Related: There is also some old issue about this in PyCo: https://github.com/pastewka/PyCo/issues/299
I think the readers should return always `None` if needed, but not `(None,)` or `(None,None)`.

**Comment by pastewka** [comment]
Okay. The only thing to keep in mind is that while the native reader may complain if `physical_sizes` is not specified, the NetCDF reader will not because `physical_sizes` is always stored in the NetCDF file. The user may in TopoBank however update the `physical_sizes` manually, but I guess it is generally good to re-read the original file in that case.

**Review by mcrot — COMMENTED** [review]
I'm also looking forward to this format, thanks. 

One question: The topography also has a flag "periodic". I've seen this is used for x_var
and y_var, but isn't this also a metadata item for the topography? On the other hand,
in TopoBank this is always entered by the user (not from file), if I remember correctly.

**Comment by pastewka** [comment]
I think I've addressed all issues. @AntoineSIMTEK can you have a look at this again?

**Review by AntoineSIMTEK — APPROVED** [review]
Looks good to me

**Pull request merged**